"""
Bulk ChIP-seq analysis pipeline.
"""

#configfile: "config.yaml"

import yaml
import sys
import os

# output directory
OUT_DIR = config["outprefix"] + "_Result"

# file prefix
tfiles = config["tfileprefix"]
cfiles = config["cfileprefix"]

# number of replicates for sample1 and sample2 conditions

## treat
n_t = len( tfiles )
tnames = [ "%s_r%d" % (config["sample"],x) for x in range( 1, n_t + 1 ) ]

## control
n_c = len( cfiles )
cnames = [ "control_r%d" % (x) for x in range( 1, n_c + 1 ) ]

# make a dictionary to match name to actual raw file prefix
t_dict_list = list( zip( tnames, tfiles ) )

sample_dict = dict( t_dict_list )

c_dict_list = list( zip( cnames, cfiles ) )

sample_dict.update( dict( c_dict_list ) )

# make links

if not os.path.exists( OUT_DIR ):
    os.mkdir( OUT_DIR )

if not os.path.exists( OUT_DIR + "/Raw" ):
    os.mkdir( OUT_DIR + "/Raw" )

for p in sample_dict:
    src = os.path.abspath(config["fastqdir"]) + "/" + sample_dict[ p ] +".fastq.gz"
    tgt = OUT_DIR + "/Raw/" + p +".fastq.gz"
    if not os.path.exists( tgt ):
        os.symlink( src, tgt )
    else:
        if os.path.realpath(src) != os.path.realpath(tgt):
            os.remove( tgt )

# individual sample
PEAKS   = expand("%s/Analysis/{sample}_peaks.narrowPeak" % (OUT_DIR), sample=tnames)
BIGWIG_RAW   = expand("%s/Analysis/{sample}_t_raw.bw" % (OUT_DIR), sample=tnames)
BIGWIG_LOGFC   = expand("%s/Analysis/{sample}_logfc.bw" % (OUT_DIR), sample=tnames)
BIGWIG_PSCORE   = expand("%s/Analysis/{sample}_pscore.bw" % (OUT_DIR), sample=tnames)
BIGWIG_SPMR   = expand("%s/Analysis/{sample}_t_spmr.bw" % (OUT_DIR), sample=tnames)
SEQ_STAT_T = expand("%s/QC/{sample}.stat.txt" % (OUT_DIR), sample=tnames)
SEQ_STAT_C = expand("%s/QC/{sample}.stat.txt" % (OUT_DIR), sample=cnames)
PEAK_STAT= expand("%s/QC/{sample}.peakstat.txt" % (OUT_DIR), sample=tnames)
PEAK_ANNO= expand("%s/Analysis/{sample}_peaks.narrowPeak_chip_annotation.pdf" % (OUT_DIR), sample=tnames)
PEAK_GIGGLE= expand("%s/Analysis/{sample}_peaks.narrowPeak_giggle_cistrome.pdf" % (OUT_DIR), sample=tnames)
#BIN_COUNT = expand("%s/Analysis/{sample}.bincount.txt" % (OUT_DIR), sample=tnames)

# combined files
SEQ_QC_SUMMARY = "%s/QC/%s.seqqc.stat.summary.txt" % (OUT_DIR, config["outprefix"])
PEAK_QC_SUMMARY = "%s/QC/%s.peakqc.stat.summary.txt" % (OUT_DIR, config["outprefix"])
#CONSENSUS = "%s/Analysis/%s.consensus.bed" % (OUT_DIR, config["outprefix"])
#BIN_CONSENSUS = "%s/Analysis/%s.binconsensus.bed" % (OUT_DIR, config["outprefix"])
#BIN_COUNT_TABLE = "%s/Analysis/%s.bincount.table" % (OUT_DIR, config["outprefix"])
GSSMAT = "%s/Analysis/%s.gss.mat.gz" % (OUT_DIR, config["outprefix"])
GSSHEATMAP = "%s/Analysis/%s.gss.heatmap.png" % (OUT_DIR, config["outprefix"])
GSSPROFILE = "%s/Analysis/%s.gss.profile.png" % (OUT_DIR, config["outprefix"])
	
TARGET = []
TARGET.extend(PEAKS + BIGWIG_RAW + BIGWIG_SPMR + BIGWIG_LOGFC + BIGWIG_PSCORE + SEQ_STAT_T + SEQ_STAT_C + PEAK_STAT + PEAK_ANNO + PEAK_GIGGLE + [SEQ_QC_SUMMARY, PEAK_QC_SUMMARY, GSSHEATMAP, GSSPROFILE] )

# decide certain parameters based on config.yaml
mapper = config["options"]["mapper"]
macs3_genome_option = " -g %s " % ( config["options"]["species"] )
macs3_cutoff_option = " -p " if config["options"]["peakcalling_cutoff_method"] == "pvalue" else " -q "
macs3_cutoff_option += str(config["options"]["peakcalling_cutoff"])
macs3_format_option = " -f BAMPE " if config["options"]["peakcalling_PE_mode"] == "Y" else " -f BAM "
macs3_lambda_option = " --nolambda " if config["options"]["peakcalling_disable_lambda"] == "Y" else " "
macs3_callpeak_option = f"--keep-dup all -B {macs3_genome_option} {macs3_cutoff_option} {macs3_lambda_option} {macs3_format_option}"
macs3_pileup_option = f" {macs3_format_option} "

# decide filtering parameter
if config["options"]["paired"] == "Y":
    bwflag_filter = " -F 3840 -f 3"
else:
    bwflag_filter = " -F 3840"

# one line to rule all

rule all:
    input: TARGET

# ---- all the following rules are common rules for analyzing a single dataset ---
include: "rules/chip_mapping.smk"
include: "rules/chip_cleanbam.smk"
include: "rules/chip_macs.smk"
#include: "rules/chip_peak_annotating.smk"
include: "rules/chip_qc.smk"
include: "rules/chip_annotate.smk"

# ---- integrative analysis of all samples ---

#include: "rules/chip_dar.smk"
#include: "rules/chip_dar_annotating.smk"

