"""
Bulk ATAC-seq analysis pipeline.
"""

configfile: "config.yaml"

import yaml
import sys
import os

ALL_SAMPLES1 = config["samples1"]
ALL_SAMPLES2 = config["samples2"]

OUT_DIR = config["outprefix"] + "_Result"

PEAKS1   = expand("%s/Analysis/{sample}_peaks.narrowPeak" % (OUT_DIR), sample=ALL_SAMPLES1)
PEAKS2   = expand("%s/Analysis/{sample}_peaks.narrowPeak" % (OUT_DIR), sample=ALL_SAMPLES2)
SEQSTAT1 = expand("%s/QC/{sample}.stat.txt" % (OUT_DIR), sample=ALL_SAMPLES1)
SEQSTAT2 = expand("%s/QC/{sample}.stat.txt" % (OUT_DIR), sample=ALL_SAMPLES2)
PEAKSTAT1= expand("%s/QC/{sample}.peakstat.txt" % (OUT_DIR), sample=ALL_SAMPLES1)
PEAKSTAT2= expand("%s/QC/{sample}.peakstat.txt" % (OUT_DIR), sample=ALL_SAMPLES2)

TARGET = []
TARGET.extend(PEAKS1 + PEAKS2 + SEQSTAT1 + SEQSTAT2 + PEAKSTAT1 + PEAKSTAT2 )

# decide certain parameters based on config.yaml
mapper = config["options"]["mapper"]
macs2_genome_option = " -g hs " if config["options"]["species"] == "GRCh38" else " -g mm "
macs2_cutoff_option = " -p " if config["options"]["peakcalling_cutoff_method"] == "pvalue" else " -q "
macs2_cutoff_option += str(config["options"]["peakcalling_cutoff"])
macs2_format_option = " -f BAMPE " if config["options"]["peakcalling_PE_mode"] == "Y" else " -f BAM "
macs2_lambda_option = " --nolambda " if config["options"]["peakcalling_disable_lambda"] == "Y" else " "
macs2_option = f"{macs2_genome_option} {macs2_cutoff_option} {macs2_lambda_option} {macs2_format_option}"

# one line to rule all

rule all:
    input: TARGET

# ---- all the following rules are common rules for analyzing a single dataset ---

include: "rules/atac_mapping.smk"
include: "rules/atac_macs2.smk"
#include: "rules/atac_peak_annotating.smk"
include: "rules/atac_qc.smk"


# ---- integrative analysis of all samples ---

#include: "rules/atac_dar.smk"
#include: "rules/atac_dar_annotating.smk"

