"""
Bulk ATAC-seq analysis pipeline.
"""

configfile: "config.yaml"

import yaml
import sys
import os

# file prefix
SAMPLE1_FILES = config["sample1fileprefix"]
SAMPLE2_FILES = config["sample2fileprefix"]

# output directory
OUT_DIR = config["outprefix"] + "_Result"

# number of replicates for sample1 and sample2 conditions
n_sample1 = len( SAMPLE1_FILES )
n_sample2 = len( SAMPLE2_FILES )

# individual sample
SAMPLE1_NAMES = expand("%s_r{i}" % config["sample1"], i = range( 1, n_sample1 + 1 ) )
SAMPLE2_NAMES = expand("%s_r{i}" % config["sample2"], i = range( 1, n_sample2 + 1 ) )
PEAKS1   = expand("%s/Analysis/{sample}_peaks.narrowPeak" % (OUT_DIR), sample=SAMPLE1_NAMES)
PEAKS2   = expand("%s/Analysis/{sample}_peaks.narrowPeak" % (OUT_DIR), sample=SAMPLE2_NAMES)
SEQSTAT1 = expand("%s/QC/{sample}.stat.txt" % (OUT_DIR), sample=SAMPLE1_NAMES)
SEQSTAT2 = expand("%s/QC/{sample}.stat.txt" % (OUT_DIR), sample=SAMPLE2_NAMES)
PEAKSTAT1= expand("%s/QC/{sample}.peakstat.txt" % (OUT_DIR), sample=SAMPLE1_NAMES)
PEAKSTAT2= expand("%s/QC/{sample}.peakstat.txt" % (OUT_DIR), sample=SAMPLE2_NAMES)
FRAGSTAT1= expand("%s/Alignment/{sample}.sortedByPos.rmdp.clean.unique.bam.insertl.txt" % (OUT_DIR), sample=SAMPLE1_NAMES)
FRAGSTAT2= expand("%s/Alignment/{sample}.sortedByPos.rmdp.clean.unique.bam.insertl.txt" % (OUT_DIR), sample=SAMPLE2_NAMES)

TARGET = []
TARGET.extend(PEAKS1 + PEAKS2 + SEQSTAT1 + SEQSTAT2 + PEAKSTAT1 + PEAKSTAT2 + FRAGSTAT1 + FRAGSTAT2 )

# decide certain parameters based on config.yaml
mapper = config["options"]["mapper"]
macs3_genome_option = " -g %s " % ( config["options"]["species"] )
macs3_cutoff_option = " -p " if config["options"]["peakcalling_cutoff_method"] == "pvalue" else " -q "
macs3_cutoff_option += str(config["options"]["peakcalling_cutoff"])
macs3_format_option = " -f BAMPE " if config["options"]["peakcalling_PE_mode"] == "Y" else " -f BAM "
macs3_lambda_option = " --nolambda " if config["options"]["peakcalling_disable_lambda"] == "Y" else " "
macs3_option = f"--keep-dup all -B --SPMR {macs3_genome_option} {macs3_cutoff_option} {macs3_lambda_option} {macs3_format_option}"

# one line to rule all

rule all:
    input: TARGET

# ---- all the following rules are common rules for analyzing a single dataset ---

include: "rules/atac_mapping.smk"
include: "rules/atac_macs.smk"
#include: "rules/atac_peak_annotating.smk"
include: "rules/atac_qc.smk"


# ---- integrative analysis of all samples ---

#include: "rules/atac_dar.smk"
#include: "rules/atac_dar_annotating.smk"

