---
title: "RNAseq Differential Expression Analysis"
author: "Department of Biostat and Bioinfo"
date: "`r Sys.Date()`"
output: 
  html_document:
    theme: readable
    toc: true
    toc_float: true
    df_print: paged
---

```{r configuration, include = FALSE}
args <- commandArgs(TRUE)

print(args)

condition1name <- args[2]
condition2name <- args[3]

metafilename <- args[4]
geneinforfilename <- as.character( args[5] ) 
MSigDBfoldername <- as.character( args[6] )
# cutoff settings
cutoff.FDR <- as.numeric( args[7] )
cutoff.log2FC <- as.numeric( args[8] )
```

```{r setup, include = FALSE}
knitr::opts_chunk$set(echo = FALSE, message = F, warning = F, dev="CairoPNG", dpi=150)

library(tximport)
library(DESeq2)
library(DEGreport)
library(ggplot2)
library(patchwork)
library(pcaExplorer)
library(BiocParallel)
library(RColorBrewer)
library(pheatmap)
library(scales)
library(ggrepel)
library(dplyr)
#library(tidyverse)
library(clusterProfiler)
library(plotly)
library(enrichplot)
library(EnhancedVolcano)
library(org.Mm.eg.db)
library(data.table)
#
library(ggnewscale)
library(MatchIt)
library(reshape2)
library(DT)
library(readxl)
library(pathview)
#
register(MulticoreParam(4))
options(stringsAsFactors=FALSE)
options(DT.warn.size=FALSE)

gseanalysis <- function (genes, gset2gene, label="", qvalueCutoff=0.2, seed=12345, minGSSize=10, maxGSSize=200 ) {
    d <- genes[,c("id","log2FoldChange")]
    d$id <- gsub("\\.\\d+","",d$id)
    rownames(d) <- d$id
    mapping <- bitr(rownames(d), fromType="ENSEMBL", toType="ENTREZID", OrgDb="org.Mm.eg.db", drop=T) 
    mapping$log2FoldChange <- d[mapping$ENSEMBL,"log2FoldChange"]
    geneList <- mapping[,"log2FoldChange"] # extract lfc
    names(geneList) <- mapping[,"ENTREZID"]
    geneList <- sort(geneList, decreasing = TRUE)
    en.gsea <- GSEA(geneList, TERM2GENE = gset2gene, verbose=FALSE, pAdjustMethod="BH", pvalueCutoff=qvalueCutoff, seed=seed,minGSSize=minGSSize, maxGSSize=maxGSSize) # main function
    en.gsea <- setReadable(en.gsea, org.Mm.eg.db, keyType = "ENTREZID")
    
    if ( length(en.gsea@result$ID) > 0 ) {
        n <- min(length(en.gsea@result$ID), 10)
    	  write.table(en.gsea, file=paste0("en.gsea.",label,".tsv"), sep = "\t")
        p <- dotplot(en.gsea, showCategory=n, font.size = 8) + ggtitle("dotplot for GSEA")
        ggsave(plot=p, file=paste0("en.gsea.",label,".dotplot.png"), height=10, width=8)
	      #p <- ridgeplot(en.gsea) + ggtitle("ridgeplot for GSEA")
	      #ggsave(plot=p, file=paste0("en.gsea.",label,".ridgeplot.png"), height=10, width=8)
        p <- gseaplot2(en.gsea, geneSetID = 1:n) + ggtitle("gseaplot for GSEA")
	      ggsave(plot=p, file=paste0("en.gsea.",label,".gseaplot.png"), height=8, width=12)
    }
    return(en.gsea)
}
```

```{r}

hgmtfile <- paste0(MSigDBfoldername,"/mh.all.v0.3.entrez.gmt")   # hallmark gene sets
h2gene <- read.gmt(hgmtfile)

cgmtfile <- paste0(MSigDBfoldername,"/m2.all.v0.3.entrez.gmt")  # Curated gene sets
c2gene <- read.gmt(cgmtfile)

ccpgmtfile <- paste0(MSigDBfoldername,"/m2.cp.v0.3.entrez.gmt")  # Curated Canonical Pathway gene sets
ccp2gene <- read.gmt(ccpgmtfile)

ogmtfile <- paste0(MSigDBfoldername,"/m4.all.v0.3.entrez.gmt")  # Oncology gene sets
o2gene <- read.gmt(ogmtfile)

tgmtfile <- paste0(MSigDBfoldername,"/m3.all.v0.3.entrez.gmt")  # Regulatory target gene sets
t2gene <- read.gmt(tgmtfile)

ggmtfile <- paste0(MSigDBfoldername,"/m5.all.v0.3.entrez.gmt")  # GO gene sets
g2gene <- read.gmt(ggmtfile)

# build meta information
meta<-read.table(metafilename,sep="\t",head=T,check.names=F) #sample,group,replicate,QC, gene, isoform

contrast.name <- paste0(condition2name, ".vs.",condition1name)

#files<-list.files(rsemfoldername,".results",recursive=T, full.names=T)
#names(files) <- sub(".results","",basename(files))

```


```{r readdata}
## Species: Mouse
## Library: mRNA

ensembl2entrez <- as.list(org.Mm.egENSEMBL2EG)
# colData
colData <- meta
rownames(colData) <- colData[,1] 
colData<-colData[,-1]
colData$Replicate <- as.factor(colData$Replicate)
colData$Group <- as.factor(colData$Group)
```


```{r readsummary}
#sum.file.path <- list.files(path=summaryfoldername, pattern = ".summary.txt", recursive = T, full.names = T)
sum.file.path <- colData$QC
length.list <- length(sum.file.path)

sum.df <- lapply(sum.file.path, 
                        function(x) (fread(x, 
                                          sep ="\t", sep2 = "|", 
                                          skip = 5, fill = T, 
                                          col.names = c("sample", gsub(".STAR.txt", "", basename(x)))) %>%
                          data.frame(row.names = 1, check.names=FALSE)) ) %>%
  do.call(what=cbind) %>% filter(!row_number() %in% c(3, 18, 23, 30)) %>%
  mutate_each_(funs(sapply(parse( text = gsub("%", "/100",.) ), eval))  , 1:length.list)

sum.Colname <- gsub(" \\|", "", rownames(sum.df))

rna.summ <- as.data.frame(t(sum.df))
colnames(rna.summ) <- sum.Colname
rna.summ2<-rna.summ[,c(4,12,13,15,seq(18,28,2))]
rna.summ2 <- rna.summ2 %>% mutate_each_(funs(as.numeric(gsub("%", "", ., fixed = T))), 1:ncol(rna.summ2)) 
```

# RNA-seq quality

This section contains QC metrics for this RNA-seq dataset.

## Sequencing depth

```{r barplotQCdepth, fig.height=5,fig.width=8}
un <- rna.summ$`Number of input reads`- rna.summ$`Uniquely mapped reads number`
m <- rna.summ$`Uniquely mapped reads number`
rname <- row.names(rna.summ)
d<-matrix()
d <- cbind("ALIGNED",m,rname)
d2 <- cbind("UNALIGNED",un,rname)
d <- rbind(d,d2)
d <- as.data.frame(d)
colnames(d) <- c("variable","value","sample")
d$value<-as.numeric(d$value)
ggplot(d,aes(x=sample,y=value,fill=variable))+geom_bar(stat="identity")+ theme(plot.title = element_text(size=11), axis.text.x = element_text(angle = 90, hjust = 1)) + xlab("") + ylab("Basepairs")+labs(fill="") + coord_flip()
```

## RNA-seq Metrics by Picard

```{r boxplotQC, fig.height=8, fig.width=12, warning=FALSE}
d <- melt(rna.summ2)
ggplot(d,aes(x=variable,y=value*100,fill=variable))+geom_boxplot(notch = F) + geom_jitter(color="black", size=2, alpha=0.9) + theme(legend.position="none", text=element_text(size=18), axis.text.x = element_text(angle = 90, hjust = 1)) + xlab("") + ylab("Percentage (%)") + coord_flip()
```

# Differential Gene Expression analysis. 

We have `r length(colData$Gene)` samples in total. We will use DESeq2
to call differentially expressed genes.

```{r deseq2run, include=FALSE}
ensembl2entrez<-as.list(org.Mm.egENSEMBL2EG)
raw <- tximport(colData$Gene, type="rsem", txIn=FALSE)
raw$length[raw$length==0] <- 1

# DESeq2
# filter no expression
dds<-DESeqDataSetFromTximport(raw, colData=colData, design=~Group)
dds<-dds[rowSums(counts(dds))>0, ]
gene.info<-read.table(geneinforfilename, sep="\t", header = T, col.names=c("id","name","class"))
gene.info$ensembl <- gsub("\\.\\d+$", "", gene.info$id)
gene.info$entrez <- as.character(ensembl2entrez[gsub("\\.\\d+$", "", gene.info$id)])
mcols(dds)<-DataFrame(mcols(dds), gene.info[match(rownames(dds),gene.info$id),])
dds<-DESeq(dds, parallel=T)

counts.data <- counts(dds, normalized=TRUE)
rld<-rlog(dds, blind=FALSE)
vsd<-varianceStabilizingTransformation(dds, blind=FALSE)
# vsd.auto<-varianceStabilizingTransformation(dds.auto,blind=FALSE)
```

## PCA plot

The Principal component analysis (PCA) provides an overview of how
different samples are similar to different to each other. This is a
good QC measurement as well since we expect to see large difference
between groups, and small difference within groups.

```{r PCA, fig.height=8, fig.width=8}
plotPCA(vsd, intgroup=c("Group")) + 
  geom_label_repel(aes(label = rownames(vsd@colData))) + 
  theme(legend.position = "none") + xlim(-15, 15) + ylim(-10,10)
```

## Correlation heatmap

Correlation heatmap shows the correlation between each pair of
samples. Similar to PCA, it can be used to measure the quality of the
data.

```{r CorHeatmap, fig.height=6,fig.width=8}
sampleDists<-dist(t(assay(vsd)))
sampleDistMatrix<-as.matrix(sampleDists)
colors<-colorRampPalette(rev(brewer.pal(9,"Blues")))(255)
pheatmap(sampleDistMatrix,clustering_distance_rows=sampleDists,clustering_distance_cols=sampleDists,col=colors,annotation_row=colData[,"Group",drop=F], show_rownames = F, fontsize_row = 6, clustering_method = "single" )
```

```{r compilealldata}
# all data will be saved into a file called `alldata.tsv`
vsd.a <- assay(vsd)
rownames(gene.info) <- gene.info$id
vsd.a.r <- rownames(vsd.a)
vsd.all <- cbind(gene.info[vsd.a.r,], vsd.a)
write.table(vsd.all, file="alldata.tsv",row.names=FALSE) # 25937 rows
```

## Master table of variance stabilized counts

```{r}
DT::datatable(vsd.all, rownames = F)
```

# DEG Analysis

The differentially expressed genes will be called through DESeq2 with
FDR cutoff `r cutoff.FDR` and minimum |log2 foldchange| of `r cutoff.log2FC`.

But first, we will show the QC of the DEG analysis using a package
called `DEGreport`. 

p-value distribution gives an idea on how well you model is capturing
the input data and as well whether it could be some problem for some
set of genes. In general, you expect to have a flat distribution with
peaks at 0 and 1. In this case, we add the mean count information to
check if any set of genes are enriched in any specific p-value range.

Variation (dispersion) and average expression relationship shouldnâ€™t
be a factor among the differentially expressed genes. When plotting
average mean and standard deviation, significant genes should be
randomly distributed.

In this case, it would be good to look at the ones that are totally
outside the expected correlation.

```{r deg_qc, ig.height=8,fig.width=12}
res<-results(dds)
design <- as.data.frame(colData(dds))
degQC(counts.data, design$Group, pvalue=res$pvalue)
```

## Differential analysis result of all the genes no matter DEG or not

This master table contains every gene before we apply the cutoffs.

```{r}
deg.res <- results(dds, contrast = c("Group", condition2name, condition1name))
m.dds <- mcols(dds)
deg.genes <- rownames(deg.res)

deg.data <- data.frame(id = deg.genes, 
                       m.dds[deg.genes, c("ensembl", "entrez","name","class")],
                       deg.res[deg.genes, c("baseMean","log2FoldChange","pvalue","padj")])
DT::datatable(deg.data, rownames = F)
```

## Differential expression gene list

The Differentially Expressed Genes are called in DESeq2, with cutoff
FDR `r cutoff.FDR` and minimum |log2 foldchange| of `r cutoff.log2FC`.

### All DEGs

```{r}
e <- deg.data %>% arrange(padj) %>% filter(padj<=cutoff.FDR) %>% filter(abs(log2FoldChange)>=cutoff.log2FC)
DT::datatable(e, rownames = F)
write.table(deg.data, file=paste0(contrast.name,".genes.all.tsv"),row.names = F)
write.table(e,  file=paste0(contrast.name,".genes.passed.tsv"),row.names = F)
```

### Up-regulated DEGs

```{r}
e.up <- e %>% filter(log2FoldChange>0)
DT::datatable(e.up, rownames = F)
```

### Down-regulated DEGs

```{r}
e.down <- e %>% filter(log2FoldChange<0)
DT::datatable(e.down, rownames = F)

write.table((e %>% filter(log2FoldChange>0))$name, 
            file=paste0(contrast.name,".deg.data.up.genenames.txt"),row.names=F,col.names = F, quote = F) 
write.table((e %>% filter(log2FoldChange<0))$name,
            file=paste0(contrast.name,".deg.data.down.genenames.txt"),row.names=F,col.names = F, quote = F) 
```

## Heatmap of DEGs

This shows the vsd data of all called DEGs, annotated with the group
information.

```{r}
pheatmap(vsd.a[e$id,], annotation_col=colData[,c("Group","Replicate"),drop=F], fontsize_col = 6, fontsize_row = 4, cluster_cols=FALSE, show_rownames=F, scale="row")
```

## MA-plot

The plot visualizes the differences between measurements taken in two
samples, by transforming the data onto M (log ratio) and A (mean
average) scales, then plotting these values. It is used to visualize
if the DEGs called are reasonable and is there any extra normalization
needed.

```{r fig.height=8, fig.width=12}
plot(log10(deg.res$baseMean),deg.res$log2FoldChange,pch=20,cex=1,ylab=paste0("log2 foldchange(",condition2name, "/", condition1name, ")"),xlab="log10(mean)",col="#BBBBBB14")
legend("bottomleft",c("Up","Down"),col=c(muted("blue"),muted("red")),pch=c(20,20))
points(log10(e.up$baseMean),e.up$log2FoldChange,col=muted("blue"),pch=20,cex=1,)
points(log10(e.down$baseMean),e.down$log2FoldChange,col=muted("red"),pch=20,cex=1,)
abline(h=0)
```

## Volcano plot

It plots significance versus fold-change on the y and x axes,
respectively. It helps deciding cutoffs and also highlight those
significantly differentiated genes. In this plot, we will mark the
top20 DEGs (in protein_coding category).

```{r fig.height=8, fig.width=12}
options(ggrepel.max.overlaps = Inf)
topgenes.for.plot <- (e %>% filter(class=="protein_coding") %>% filter(padj<=1e-10) %>% 
                        top_n(20, wt = abs(log2FoldChange) ))$name

EnhancedVolcano(deg.data,
                 lab = deg.data$name,
                 x = 'log2FoldChange',
                 y = 'padj',
                 selectLab = topgenes.for.plot,        #selected_genes$geneSymbol,
                 xlab = bquote(~Log[2]~ 'fold change'),
                 ylab = bquote(~-Log[10]~ 'FDR'),
                 pCutoff = cutoff.FDR,
                 FCcutoff = cutoff.log2FC,
                 pointSize = 2.0,
                 labSize = 3.0,
                 labCol = 'black',
                 labFace = 'bold',
                 boxedLabels = FALSE,
                 colAlpha = 0.5,
                 legendPosition = 'right',
                 legendLabSize = 10,
                 legendIconSize = 4.0,
                 drawConnectors = TRUE,
                 widthConnectors = 0.2,
                 colConnectors = 'black') + xlim(-15,15) + theme(axis.text.x=element_text(size=10,face="bold"), axis.text.y=element_text(size=10,face="bold"), axis.title=element_text(size=12,face="bold"))
```

# Gene set enrichment analysis using MSigDB

We will use the gene sets from MSigDB mouse (DRAFT). Here are the sets we will check:

1. MH: mouse-ortholog hallmark gene sets

   These are versions of gene sets in the MSigDB Hallmark collection
   mapped to their Mouse orthologs.

   Note: Gene sets in MH consist of human genes converted to their mouse
   orthologs and are not originally sourced from mouse datasets.

2. M2: curated gene sets

   Gene sets in this collection are curated from various sources,
   including the Reactome, WikiPathways, and BioCarta pathway databases
   and also contributed by individual domain experts.

   1. M2:CP Canonical pathways

      Gene sets from pathway databases. Usually, these gene sets are
      canonical representations of a biological process compiled by
      domain experts.

      These gene sets consist solely of natively mouse gene sets, and
      are generated through a procedure identical to their Human
      MSigDB equivalents.

3. M3: regulatory target gene sets

   Gene sets representing potential targets of regulation by
   transcription factors or microRNAs. The sets consist of genes
   grouped by elements they share in their non-protein coding
   regions. The elements represent known or likely cis-regulatory
   elements in promoters and 3'-UTRs. The C3 collection is divided
   into two sub-collections: microRNA targets from miRDB and
   transcription factor targets from GTRD.  These gene sets consist
   solely of natively mouse gene sets.

4. M4: Tumor ontology sets

   Gene sets from the MP Tumor ontology resource at MGI.

5. M5: GO gene sets

   Mouse gene sets from the Gene Ontology Consortium

   These gene sets consist solely of natively mouse gene sets, and are
   generated through a procedure identical to their Human MSigDB
   equivalents.

First, we will use GSEA (gene-set enrichment analysis) to ask which
gene set is more related to the differential gene expression pattern
we observed in the experiment. As for GSEA, all genes will be ranked
by their log foldchange from high to low (or upregulated to
downregulated). We will use adjusted-pvalue of 0.5 as cutoff and show
at most top 5 gene sets for each of the six categories of genes sets.

```{r}
p.hallmark <- gseanalysis( deg.data, h2gene, paste0( contrast.name,".hallmark" ), qvalueCutoff = 0.5 )
p.curated  <- gseanalysis( deg.data, c2gene, paste0( contrast.name,".curated" ), qvalueCutoff = 0.5 )
p.canonicalpathways  <- gseanalysis( deg.data, ccp2gene, paste0( contrast.name,".canonicalpathways" ), qvalueCutoff = 0.5 )
p.regulatory  <- gseanalysis( deg.data, t2gene, paste0( contrast.name,".regulatory" ), qvalueCutoff = 0.5 )
p.oncology <- gseanalysis( deg.data, o2gene, paste0( contrast.name,".oncology" ), qvalueCutoff = 0.5 )
p.go       <- gseanalysis( deg.data, g2gene, paste0( contrast.name,".go" ), qvalueCutoff = 0.5 )

```

Note: if there is no figure, it means there is no enriched gene sets
found under this category.

## GSEA Hallmark gene sets

```{r fig.height=6, fig.width=12}
n.p.hallmark <- ifelse(!is.null(p.hallmark), length(p.hallmark@result$ID), 0 )

if ( n.p.hallmark > 0 ) {
  n <- min(n.p.hallmark, 5) 
  gseaplot2(p.hallmark, geneSetID = 1:n, pvalue_table = T) + ggtitle(paste0("GSEA: Top ", as.character(n), " hallmarks"))
}
```

## GSEA Curated gene sets

```{r fig.height=6, fig.width=12}
n.p.curated <- ifelse(!is.null(p.curated), length(p.curated@result$ID), 0 )

if ( n.p.curated > 0 ) {
  n <- min(n.p.curated, 5) 
  gseaplot2(p.curated, geneSetID = 1:n, pvalue_table = T) + ggtitle(paste0("GSEA: Top ", as.character(n), " curated"))
}
```

## GSEA Canonical Pathways

```{r fig.height=6, fig.width=12}
n.p.canonicalpathways <- ifelse(!is.null(p.canonicalpathways), length(p.canonicalpathways@result$ID), 0 )

if ( n.p.canonicalpathways > 0 ) {
  n <- min(n.p.canonicalpathways, 5)
  gseaplot2(p.canonicalpathways, geneSetID = 1:n, pvalue_table = T) + ggtitle(paste0("GSEA: Top ", as.character(n), " canonical pathways"))
}
```

## GSEA Regulatory Targets

```{r fig.height=6, fig.width=12}
n.p.regulatory <- ifelse(!is.null(p.regulatory), length(p.regulatory@result$ID), 0 )

if ( n.p.regulatory > 0 ) {
  n <- min(n.p.regulatory, 5) 
  gseaplot2(p.regulatory, geneSetID = 1:n, pvalue_table = T) + ggtitle(paste0("GSEA: Top ", as.character(n), " canonical pathways"))
}
```

## GSEA Tumor Ontology gene sets

```{r fig.height=6, fig.width=12}
n.p.oncology <- ifelse(!is.null(p.oncology), length(p.oncology@result$ID), 0 )

if ( n.p.oncology > 0 ) {
  n <- min(n.p.oncology, 5) 
  gseaplot2(p.oncology, geneSetID = 1:n, pvalue_table = T) + ggtitle(paste0("GSEA: Top ", as.character(n), " oncology"))
}
```

## GSEA GO gene sets

For better GO term analysis, please check OR analysis based on GO
terms.

```{r fig.height=6, fig.width=12}
n.p.go <- ifelse(!is.null(p.go), length(p.go@result$ID), 0 )

if ( n.p.go > 0 ) {
  n <- min(n.p.go, 5) 
  gseaplot2(p.go, geneSetID = 1:n, pvalue_table = T) + ggtitle(paste0("GSEA: Top ", as.character(n), " GO"))
}
```

# Gene function Over-Representation analysis 

As for OR analysis, we are only focusing on the called 'differentially
expressed genes' or DEGs. Then a fisher test will be performed for
each 'term' (or the name of a gene set) to see if the term and the
DEGs are associated. We will search for all possible enriched GO terms
associated with DEGs, then perform a semantic analysis to simplfy and
cluster the result. To make comparisons fair, we generated a 'gene
universe' or the 'background' gene set to match the DEGs. We selected
those non-differentially expressed genes which have similar gene
expression levels as DEGs. We repeat this process five times so that
the 'background' pool can be much larger than DEGs. In the OR
analysis, we will use non-DEGs + DEGs as the 'gene universe'.

```{r make_bg}
backM <- c()

## prepare data frame for matching, sign indicates wheather
## the gene is differentially expressed or not
df <- data.frame( sign=as.numeric( rownames(deg.res)  %in% as.character(rownames(e))),res["baseMean"])
df$baseMean <- round( df$baseMean, 0)

## repeat matching multiple times since
## each differentially expressed gene is
## matched by exactly one non-expressed gene

for( i in 1:5 ){
  mm <- matchit( sign ~ baseMean, df, method="nearest", distance="mahalanobis")
  backM <- c(backM, mm$match.matrix[,1])
  df <- df[which(!rownames(df) %in% backM),]
}

backM <- unique( na.exclude(backM) )
backM.data <- deg.data[backM, c("ensembl","entrez","name","log2FoldChange")]

id.all <- e$log2FoldChange
names(id.all) <- e$ensembl
id.all.ensembl <- names(id.all)
id.all.entrez <- bitr(id.all.ensembl, fromType="ENSEMBL", toType=c("ENTREZID"), OrgDb="org.Mm.eg.db")$ENTREZ

id.up <- e.up$log2FoldChange
names(id.up) <- e.up$ensembl
id.up.ensembl <- names(id.up)
id.up.entrez <- bitr(id.up.ensembl, fromType="ENSEMBL", toType=c("ENTREZID"), OrgDb="org.Mm.eg.db")$ENTREZ

id.down <- e.down$log2FoldChange
names(id.down) <- e.down$ensembl
id.down.ensembl <- names(id.down)
id.down.entrez <- bitr(id.down.ensembl, fromType="ENSEMBL", toType=c("ENTREZID"), OrgDb="org.Mm.eg.db")$ENTREZ
```

## GO term analysis for all differentially expressed genes

*Enriched terms*

```{r cp0}
go.all <- enrichGO(id.all.ensembl, c(id.all.ensembl, backM.data$ensembl),OrgDb = org.Mm.eg.db, keyType = "ENSEMBL", ont="BP", pAdjustMethod = "BH", pvalueCutoff = 0.05, qvalueCutoff = 0.2, readable=T)
if ( dim(head(go.all))[1] >0 ) {
  go.all <- pairwise_termsim(go.all)
  go.all <- simplify(go.all, cutoff=0.7, by="p.adjust", select_fun=min)
  DT::datatable(head(go.all,10)[,c("ID","Description","GeneRatio","BgRatio","pvalue","qvalue")], caption="Top 10 GO:BP terms by ClusterProfiler")
}
```

*Treeplot* Please right-click and show the figure in seperate tab to see detail if the font is too small.

```{r cp0.draw, fig.height=12, fig.width=16}
if ( dim(head(go.all))[1] >0 ) {
  p1 <- treeplot(go.all)
  p2 <- mutate(go.all, qscore = -log(p.adjust, base=10)) %>% 
               barplot(x="qscore")
  p1 / p2
}
```

## GO term analysis for upregulated genes

*Enriched terms*

```{r cp1}
go.up <- enrichGO(id.up.ensembl, c(id.up.ensembl, backM.data$ensembl),OrgDb = org.Mm.eg.db, keyType = "ENSEMBL", ont="BP", pAdjustMethod = "BH", pvalueCutoff = 0.05, qvalueCutoff = 0.2, readable=T)
if ( dim(head(go.up))[1] >0 ) {
  go.up <- pairwise_termsim(go.up)
  go.up <- simplify(go.up, cutoff=0.7, by="p.adjust", select_fun=min)
  DT::datatable(head(go.up,10)[,c("ID","Description","GeneRatio","BgRatio","pvalue","qvalue")], caption="Top 10 GO:BP terms by ClusterProfiler")
}
```

*Treeplot* Please right-click and show the figure in seperate tab to see detail if the font is too small.

```{r cp1.draw, fig.height=12, fig.width=16}
if ( dim(head(go.up))[1] >0 ) {
  p1 <- treeplot(go.up)
  p2 <- mutate(go.up, qscore = -log(p.adjust, base=10)) %>% 
               barplot(x="qscore")
  p1 / p2
}
```

## GO term analysis for downregulated genes

*Enriched terms*

```{r cp2}
go.down <- enrichGO(id.down.ensembl, c(id.down.ensembl, backM.data$ensembl),OrgDb = org.Mm.eg.db, keyType = "ENSEMBL", ont="BP", pAdjustMethod = "BH", pvalueCutoff = 0.05, qvalueCutoff = 0.2, readable=T)
if ( dim(head(go.down))[1] >0 ) {
  go.down <- pairwise_termsim(go.down)
  go.down <- simplify(go.down, cutoff=0.7, by="p.adjust", select_fun=min)
  DT::datatable(head(go.down,10)[,c("ID","Description","GeneRatio","BgRatio","pvalue","qvalue")], caption="Top 10 GO:BP terms by ClusterProfiler")
}
```

*Treeplot* Please right-click and show the figure in seperate tab to see detail if the font is too small.

```{r cp2.draw, fig.height=12, fig.width=16}
if ( dim(head(go.down))[1] >0 ) {
  p1 <- treeplot(go.down)
  p2 <- mutate(go.down, qscore = -log(p.adjust, base=10)) %>% 
               barplot(x="qscore")
  p1 / p2
}
```

# KEGG analysis

This analysis is use OR analysis on DEGs to find the most enriched
KEGG pathway. If there is no table followed, there is no KEGG pathway
enriched.

## KEGG analysis for all differentially expressed genes

```{r kegg0}
kegg.all <- enrichKEGG(id.all.entrez, c(id.all.entrez, backM.data$entrez),organism="mmu", pAdjustMethod = "BH", pvalueCutoff = 0.05, qvalueCutoff = 0.2, keyType="kegg")
if ( length(kegg.all$ID) > 0 ) {
  DT::datatable(head(kegg.all,10)[,c("ID","Description","GeneRatio","BgRatio","pvalue","qvalue")], caption="Top 10 KEGG pathways by ClusterProfiler")
}
```

## KEGG pathway illustrations

KEGG pathway plot is overlayed with gene log2 foldchange values from
our study. In the following figures, all genes will be colored,
although the KEGG pathways are found by using only the DEGs. In
another word, some genes with great foldchange are not called as DEGs
probably due to in-significant p-value. 

For some unknown reason, the KEGG figures may not be embeded in the
HTML report. In this case, please contact us to get the figures.

```{r kegg0.draw, fig.width=16}
v <- deg.data$log2FoldChange
names(v) <- deg.data$ensembl
for ( keggid in kegg.all$ID ) {
  keggpathview <- pathview(gene.data  = v, pathway.id = keggid, gene.idtype="ENSEMBL", species    = "mmu" , limit=list(gene=2, cpd=1) )
  keggpathview.filename <- paste0("./",keggid,".pathview.png")
  knitr::include_graphics(keggpathview.filename) # this code somehow doesn't work
}
# Here we should try to run the above commands then decide which figures to include in knitr.
#knitr::include_graphics("./mmu04061.pathview.png")
#knitr::include_graphics("./mmu05146.pathview.png")
#knitr::include_graphics("./mmu05200.pathview.png")
```
