---
title: "ChIP-seq Analysis Report" 
author: "Department of Biostat and Bioinfo"
date: "`r Sys.Date()`"
output: 
  html_document:
    theme: readable
    toc: true
    toc_float: true
    df_print: paged
params:
    date: !r Sys.Date()
    name: "pipeline.name"
    peakfile: "filename"
    peakqc: "filename"
    seqqc: "filename"
    gssheatmap: "filename"
    gssprofile: "filename"
    gmtfolder: "foldername"
    cistromedb: "foldername"
---

```{r configuration, include = FALSE, echo=FALSE }
knitr::opts_chunk$set( echo = FALSE, message = FALSE, dev = "CairoPNG", dpi = 150, warning=FALSE, error=FALSE )

# input file
pipeline.name <- params$name

peakqc.file <- params$peakqc
seqqc.file <- params$seqqc
gss.heatmap <- params$gssheatmap
gss.profile <- params$gssprofile

peak.file <- params$peakfile

# msigdb
gmtfolder <- params$gmtfolder

# cistromedb
cistrome.gigglepath <- as.character( params$cistrome )
```

```{r setup, include=FALSE}
# settings
library(ggplot2)
library(RColorBrewer)
library(scales)
library(ggrepel)
library(plotly)
library(org.Hs.eg.db)
library(TxDb.Hsapiens.UCSC.hg38.knownGene)
library(ChIPseeker)
library(clusterProfiler)
library(patchwork)
library(DT)

options(stringsAsFactors=F)

# some functions
# giggle function to run against cistromeDB
RunGiggle.Cistrome <- function(peakbed, giggle.path, organism){
  antFile <- read.csv( paste0(giggle.path,"/CistromeDB.sample.annotation.txt"), sep="\t", row.names=1, stringsAsFactors = FALSE )
  outputBed = peakbed
  cmd <- paste0("sort --buffer-size 2G -k1,1 -k2,2n -k3,3n ", outputBed, " | cut -f 1,2,3 | bgzip -c > ", outputBed, ".bed.gz")
  system(cmd)
  cmd <- paste0("giggle search -g 2700000000 -i ", giggle.path, "/giggle.", organism, " -q ", outputBed, ".bed.gz -s > ", outputBed, ".result.xls")
  system(cmd)
  resultDf <- read.table(paste0(outputBed, ".result.xls"), sep="\t", row.names=NULL, comment.char="", stringsAsFactors =  FALSE)
  resultDf <- resultDf[,-9]
  colnames(resultDf) <- c("file", "file_size", "overlaps", "odds_ratio", "fishers_two_tail", "fishers_left_tail", "fishers_right_tail", "combo_score")
  resultDf <- resultDf[resultDf$overlaps>0,] #at least some overlap
  resultDf <- resultDf[resultDf$"odds_ratio">1,] # at least some enrichment over random
  rownames(resultDf) <- sapply(strsplit(resultDf$file, "/"), function(x) return(gsub("_5foldPeak.bed.gz", "", rev(x)[1])))
  resultDf <- resultDf[,c("file_size", "overlaps", "combo_score", "odds_ratio", "fishers_right_tail")]
  targetDf <- merge(resultDf, antFile, by.x=0, by.y=0)
  colnames(targetDf) <- c("sample_id", "sample_peak_number", "overlap_peak_number", "giggle_score", "odds_ratio", "fisher_right_tail", "GSM_id", "species", "factor", "cell_line", "cell_type", "tissue", "disease")

  targetDf$biological_resource <- apply(targetDf, 1, function(x) return(paste0(x[7:10], collapse=";")))
  targetDf$labels <- paste0(targetDf$"cell_line",":",targetDf$"factor")
  targetDf$fisher_right_tail <- -log10(targetDf$"fisher_right_tail")
  targetDf$"odds_ratio" <- log10(targetDf$"odds_ratio")  
  
  targetDf <- targetDf[, c("sample_id", "GSM_id", "species", "factor", "biological_resource", "labels", "giggle_score", "odds_ratio", "fisher_right_tail", "sample_peak_number", "overlap_peak_number")]
  targetDf <- targetDf[order(-targetDf$giggle_score), ]
  targetDf <- targetDf[!duplicated(targetDf$factor), ]
  write.table(targetDf, paste0(outputBed, ".giggle.res.tfs.txt"), sep="\t", quote=FALSE, row.names=FALSE)
  cmd <- paste0("rm ", outputBed, ".bed.gz")
  system(cmd)
  cmd <- paste0("rm ", outputBed, ".result.xls")
  system(cmd)
  return(targetDf)
}

# giggle function to run on ENCODE cCRE (human)
RunGiggle.cCRE <- function(peakbed, giggle.ccre.path){
  outputBed = peakbed
  # bgzip
  cmd <- paste0("sort --buffer-size 2G -k1,1 -k2,2n -k3,3n ", outputBed, " | cut -f 1,2,3 | bgzip -c > ", outputBed, ".bed.gz")
  system(cmd)
  
  # get overlaps in BED
  cmd <- paste0("giggle search -g 2700000000 -i ", giggle.ccre.path, " -q ", outputBed, ".bed.gz  -f pELS -o -v | perl -e '$_=<>;chomp;s/^##//;$p=$_;@id=();while(<>){if(/^##/){if(@id){print $p,\"\\t\",join(\",\",@id),\"\\n\";}chomp;s/^##//;$p=$_;@id=();}else{@_=split;push @id,$_[3];}}if(@id){print $p,\"\\t\",join(\",\",@id),\"\\n\";}' > ", outputBed, ".result.pELS.bed" )
  system(cmd)
  cmd <- paste0("giggle search -g 2700000000 -i ", giggle.ccre.path, " -q ", outputBed, ".bed.gz  -f dELS -o -v | perl -e '$_=<>;chomp;s/^##//;$p=$_;@id=();while(<>){if(/^##/){if(@id){print $p,\"\\t\",join(\",\",@id),\"\\n\";}chomp;s/^##//;$p=$_;@id=();}else{@_=split;push @id,$_[3];}}if(@id){print $p,\"\\t\",join(\",\",@id),\"\\n\";}' > ", outputBed, ".result.dELS.bed" )
  system(cmd)
  
  # get summary
  cmd <- paste0("giggle search -g 2700000000 -i ", giggle.ccre.path, " -q ", outputBed, ".bed.gz -s > ", outputBed, ".result.tsv")
  system(cmd)
  resultDf <- read.table(paste0(outputBed, ".result.tsv"), sep="\t", row.names=, comment.char="#", stringsAsFactors =  FALSE)
  resultDf <- resultDf[,-9]
  colnames(resultDf) <- c("class","file_size","overlaps","odds_ratio","fishers_two_tail","fishers_left_tail","fishers_right_tail","combo_score")
  
  cmd <- paste0("rm ", outputBed, ".bed.gz")
  system(cmd)
  cmd <- paste0("rm ", outputBed, ".result.tsv")
  system(cmd)
  resultDf$class <- gsub("encodeCcre/encodeCcre-","",resultDf$class)
  resultDf$class <- gsub(".bed.gz","",resultDf$class)
  return(resultDf)
}

```

```{r config.exp, include=FALSE}
# distance cutoff for chipseeker to link dar to gene
peak2gene.distance.cutoff <- 10000

# parameters for enrichr/clusterprofiler
cp.pvalueCutoff = 1.0
cp.pAdjustMethod = "BH"
cp.minGSSize = 5
cp.maxGSSize = 500
cp.qvalueCutoff = 1.0
```

```{r loadmsigdb, include=FALSE}
# load gmt files for enricher/clusterprofiler
# hall mark
hgmtfile <- paste0(gmtfolder, "/h.all.v7.5.1.entrez.gmt")
hgene <- read.gmt(hgmtfile)

# ontology
c5gmtfile <- paste0(gmtfolder, "/c5.go.bp.v7.5.1.entrez.gmt")
bpgene <- read.gmt(c5gmtfile)

# Human Phenotype Ontology
hpogmtfile <- paste0(gmtfolder, "/c5.hpo.v7.5.1.entrez.gmt")
hpogene <- read.gmt(hpogmtfile)

# oncogenic signature
c6gmtfile <- paste0(gmtfolder, "/c6.all.v7.5.1.entrez.gmt")
oncogene <- read.gmt(c6gmtfile)

# immuneSigDB
c7gmtfile <- paste0(gmtfolder, "/c7.immunesigdb.v7.5.1.entrez.gmt")
imgene <- read.gmt(c7gmtfile)

# TF targets Signatures
c3gmtfile <- paste0(gmtfolder, "/c3.tft.v7.5.1.entrez.gmt")
tftgene <- read.gmt(c3gmtfile)

# Curated Reactome Pathways
rpgmtfile <- paste0(gmtfolder, "/c2.cp.reactome.v7.5.1.entrez.gmt")
rpgene <- read.gmt(rpgmtfile)

# Curated Wikipathways 
wpgmtfile <- paste0(gmtfolder, "/c2.cp.wikipathways.v7.5.1.entrez.gmt")
wpgene <- read.gmt(wpgmtfile)

# Curated Biocarta
bcgmtfile <- paste0(gmtfolder, "/c2.cp.biocarta.v7.5.1.entrez.gmt")
bcgene <- read.gmt(bcgmtfile)

# Curated KEGG
kegggmtfile <- paste0(gmtfolder, "/c2.cp.kegg.v7.5.1.entrez.gmt")
kegggene <- read.gmt(kegggmtfile)

# Curated Canonical Pathways
cpgmtfile <- paste0(gmtfolder, "/c2.cp.v7.5.1.entrez.gmt")
cpgene <- read.gmt(cpgmtfile)
```


```{r loaddata, include=FALSE}
peaks <- readPeakFile(peak.file)
```

# Intro

# Summary of Data Quality

This section contains QC metrics for this ChIP-seq dataset.

## Sequencing QC

```{r readsummary}
seqqc <- read.table(seqqc.file, header=T)

colnames(seqqc) <- c("sample name", "replicate", "total reads", "mapped reads", "ratio mapped over total", "duplicated reads",  "ratio duplicated over total", "valid reads", "ratio valid over total", "valid in promoter", "ratio promoter over valid", "valid in peaks", "ratio peak over valid" )

peakqc <- read.table(peakqc.file, header=T)
colnames(peakqc) <- c("sample name", "replicate", "total pairs", "total peaks", "peaks in blacklist", "peaks in promoters", "ratio of peaks in promoters", "peaks in DHSs", "ratio of peaks in DHSs")

```

```{r showsqctable}
DT::datatable(seqqc)
```

## Peakcalling QC

```{r showpqctable}
DT::datatable(peakqc)
```

## ChIP signal distribution around gene start sites
```{r chipqcsignal}
knitr::include_graphics( c( gss.heatmap, gss.profile ) )
```

# Annotation of peaks

## Genomic distributions of peaks

Peaks over the whole genome:

```{r dev="png",dpi=300,fig.height=8,fig.width=8,echo=F,message=F }
p <- covplot(peaks, title="Peaks over Chromosomes") + theme(axis.text.y = element_blank(), axis.ticks.y = element_blank(), axis.line.y = element_blank())
p
```

## Genomic Features at peaks

```{r annotatepeaks, fig.height=8, fig.width=8}
peakAnno <- annotatePeak( peaks, TxDb=TxDb.Hsapiens.UCSC.hg38.knownGene, annoDb="org.Hs.eg.db", tssRegion=c(-3000, 3000), verbose=FALSE )
p1 <- plotAnnoBar(peakAnno)
p2 <- plotDistToTSS(peakAnno)
p1 / p2
```

## Cistrome Searches

Next, we aim to find the potential enriched TFs within those peaks by querying all the public available human transcription factor ChIP-seq dataset collection in Cistrome DB. For each ChIP-seq dataset, we ask how many our peaks overlap with the top1000 peaks of that dataset, then calculate a 'Giggle combo score' which is the product of the -log10 pvalue and log2 odds-ratio. 

```{r giggle.analysis}
giggle.res <- RunGiggle.Cistrome( peak.file, cistrome.gigglepath, "GRCh38")
```

There are `r length(rownames(giggle.res))` Cistromes enriched (with at least 1 peak overlap and oddsratio>1) in our peaks. The complete list are shown below:

```{r echo=FALSE}
DT::datatable(giggle.res[order(-giggle.res$giggle_score),])
```

We can plot the odds-ratio in x-axis and pvalue (Fisher-test right tail test) in y-axis, while highlighting the top 10 Cistrome hits :

```{r dev="png",dpi=300,fig.height=8,fig.width=8,echo=F,message=F }
giggle.res.top10 <- giggle.res[order(giggle.res$giggle_score,decreasing = T),][1:10,]

t <- paste0(giggle.res.top10$biological_resource,collapse = "\n")
t <- paste0("Top10 result:\n",t)

p <- ggplot(data=giggle.res, aes(x=odds_ratio, y=fisher_right_tail, label=labels)) + geom_point() + xlim(0,4) + ylim(0,250)
p<-p + geom_label_repel(force=20, size=2, data = giggle.res.top10, max.overlaps = Inf) + xlab("Giggle log10 odds_ratio") + ylab("Giggle Fisher-test right-tail -log10 pvalue") + annotate(geom = 'text', label = t, x = Inf, y = -Inf, hjust = 1., vjust = -0.5)
p

```

## Associated genes with peaks

We use distance `r peak2gene.distance.cutoff` basepairs as cutoff to define the 'nearby' genes to peaks.

```{r genespeak, fig.height=8, fig.width=8}
# define nearby genes

# extract peak2gene.distance.cutoff bps u/d GSS
gss.flanking <- trim(promoters(genes(TxDb.Hsapiens.UCSC.hg38.knownGene), upstream=peak2gene.distance.cutoff, downstream=peak2gene.distance.cutoff))

# overlap
over.gss.flanking.peaks <- findOverlaps(gss.flanking, peaks, ignore.strand=T)

# get entrez ids
over.gss.flanking.peaks.entrez <- unique(gss.flanking[over.gss.flanking.peaks@from]$gene_id)

# convert entrez to symbol
over.gss.flanking.peaks.symbol <- unique(bitr(over.gss.flanking.peaks.entrez, toType="SYMBOL", fromType="ENTREZID", OrgDb="org.Hs.eg.db")$SYMBOL)

write.table( over.gss.flanking.peaks.symbol, file=paste0(pipeline.name,"_nearby_genes.tsv") )
```

We can find `r length(over.gss.flanking.peaks.entrez)` genes near peaks.

Now we can ask what kind of gene functions they have by doing enrichment analysis (ORA).

```{r echo=FALSE,dev="png",dpi=300,fig.height=8,fig.width=16,echo=F,message=F}
ER <- enricher( over.gss.flanking.peaks.entrez, TERM2GENE = hgene, pvalueCutoff  = cp.pvalueCutoff, pAdjustMethod = cp.pAdjustMethod, minGSSize = cp.minGSSize, maxGSSize = cp.maxGSSize, qvalueCutoff = cp.qvalueCutoff )

n <- "Hallmarks"
if ( !is.null(ER) && nrow(ER) > 0 ) {
    dotplot(ER, title=n)
}
if ( !is.null(ER) && nrow(ER) > 0 ) {
    datatable(as.data.frame(ER), rownames=F)
    write.table( as.data.frame(ER), file=paste0(pipeline.name,"_",n,".tsv") )
}
```

```{r echo=FALSE,dev="png",dpi=300,fig.height=8,fig.width=16,echo=F,message=F}
ER <- enricher( over.gss.flanking.peaks.entrez, TERM2GENE = bpgene, pvalueCutoff  = cp.pvalueCutoff, pAdjustMethod = cp.pAdjustMethod, minGSSize = cp.minGSSize, maxGSSize = cp.maxGSSize, qvalueCutoff = cp.qvalueCutoff )

n <- "GP-BP"
if ( !is.null(ER) && nrow(ER) > 0 ) {
    dotplot(ER, title=n)
}
if ( !is.null(ER) && nrow(ER) > 0 ) {
    datatable(as.data.frame(ER), rownames=F)
    write.table( as.data.frame(ER), file=paste0(pipeline.name,"_",n,".tsv") )
}
```

```{r echo=FALSE,dev="png",dpi=300,fig.height=8,fig.width=16,echo=F,message=F}
ER <- enricher( over.gss.flanking.peaks.entrez, TERM2GENE = hpogene, pvalueCutoff  = cp.pvalueCutoff, pAdjustMethod = cp.pAdjustMethod, minGSSize = cp.minGSSize, maxGSSize = cp.maxGSSize, qvalueCutoff = cp.qvalueCutoff )

n <- "Human-Phenotype-Ontology"
if ( !is.null(ER) && nrow(ER) > 0 ) {
    dotplot(ER, title=n)
}
if ( !is.null(ER) && nrow(ER) > 0 ) {
    datatable(as.data.frame(ER), rownames=F)
    write.table( as.data.frame(ER), file=paste0(pipeline.name,"_",n,".tsv") )
}
```

```{r echo=FALSE,dev="png",dpi=300,fig.height=8,fig.width=16,echo=F,message=F}
ER <- enricher( over.gss.flanking.peaks.entrez, TERM2GENE = oncogene, pvalueCutoff  = cp.pvalueCutoff, pAdjustMethod = cp.pAdjustMethod, minGSSize = cp.minGSSize, maxGSSize = cp.maxGSSize, qvalueCutoff = cp.qvalueCutoff )
if ( !is.null(ER) && nrow(ER) > 0 ) {
    dotplot(ER, title="Oncogenic Signature")
}

n <- "Oncogenic-Signature"
if ( !is.null(ER) && nrow(ER) > 0 ) {
    dotplot(ER, title=n)
}
if ( !is.null(ER) && nrow(ER) > 0 ) {
    datatable(as.data.frame(ER), rownames=F)
    write.table( as.data.frame(ER), file=paste0(pipeline.name,"_",n,".tsv") )
}
```

```{r echo=FALSE,dev="png",dpi=300,fig.height=8,fig.width=16,echo=F,message=F}
ER <- enricher( over.gss.flanking.peaks.entrez, TERM2GENE = imgene, pvalueCutoff  = cp.pvalueCutoff, pAdjustMethod = cp.pAdjustMethod, minGSSize = cp.minGSSize, maxGSSize = cp.maxGSSize, qvalueCutoff = cp.qvalueCutoff )

n <- "immuneSigDB"
if ( !is.null(ER) && nrow(ER) > 0 ) {
    dotplot(ER, title=n)
}
if ( !is.null(ER) && nrow(ER) > 0 ) {
    datatable(as.data.frame(ER), rownames=F)
    write.table( as.data.frame(ER), file=paste0(pipeline.name,"_",n,".tsv") )
}
```

```{r echo=FALSE,dev="png",dpi=300,fig.height=8,fig.width=16,echo=F,message=F}
ER <- enricher( over.gss.flanking.peaks.entrez, TERM2GENE = tftgene, pvalueCutoff  = cp.pvalueCutoff, pAdjustMethod = cp.pAdjustMethod, minGSSize = cp.minGSSize, maxGSSize = cp.maxGSSize, qvalueCutoff = cp.qvalueCutoff )

n <- "TF-targets-Signatures"
if ( !is.null(ER) && nrow(ER) > 0 ) {
    dotplot(ER, title=n)
}
if ( !is.null(ER) && nrow(ER) > 0 ) {
    datatable(as.data.frame(ER), rownames=F)
    write.table( as.data.frame(ER), file=paste0(pipeline.name,"_",n,".tsv") )
}
```

```{r echo=FALSE,dev="png",dpi=300,fig.height=8,fig.width=16,echo=F,message=F}
ER <- enricher( over.gss.flanking.peaks.entrez, TERM2GENE = rpgene, pvalueCutoff  = cp.pvalueCutoff, pAdjustMethod = cp.pAdjustMethod, minGSSize = cp.minGSSize, maxGSSize = cp.maxGSSize, qvalueCutoff = cp.qvalueCutoff )

n <- "Curated-Reactome-Pathways"
if ( !is.null(ER) && nrow(ER) > 0 ) {
    dotplot(ER, title=n)
}
if ( !is.null(ER) && nrow(ER) > 0 ) {
    datatable(as.data.frame(ER), rownames=F)
    write.table( as.data.frame(ER), file=paste0(pipeline.name,"_",n,".tsv") )
}
```

```{r echo=FALSE,dev="png",dpi=300,fig.height=8,fig.width=16,echo=F,message=F}
ER <- enricher( over.gss.flanking.peaks.entrez, TERM2GENE = wpgene, pvalueCutoff  = cp.pvalueCutoff, pAdjustMethod = cp.pAdjustMethod, minGSSize = cp.minGSSize, maxGSSize = cp.maxGSSize, qvalueCutoff = cp.qvalueCutoff )

n <- "Curated-Wikipathways"
if ( !is.null(ER) && nrow(ER) > 0 ) {
    dotplot(ER, title=n)
}
if ( !is.null(ER) && nrow(ER) > 0 ) {
    datatable(as.data.frame(ER), rownames=F)
    write.table( as.data.frame(ER), file=paste0(pipeline.name,"_",n,".tsv") )
}
```

```{r echo=FALSE,dev="png",dpi=300,fig.height=8,fig.width=16,echo=F,message=F}
ER <- enricher( over.gss.flanking.peaks.entrez, TERM2GENE = bcgene, pvalueCutoff  = cp.pvalueCutoff, pAdjustMethod = cp.pAdjustMethod, minGSSize = cp.minGSSize, maxGSSize = cp.maxGSSize, qvalueCutoff = cp.qvalueCutoff )

n <- "Curated-Biocarta"
if ( !is.null(ER) && nrow(ER) > 0 ) {
    dotplot(ER, title=n)
}
if ( !is.null(ER) && nrow(ER) > 0 ) {
    datatable(as.data.frame(ER), rownames=F)
    write.table( as.data.frame(ER), file=paste0(pipeline.name,"_",n,".tsv") )
}
```

```{r echo=FALSE,dev="png",dpi=300,fig.height=8,fig.width=16,echo=F,message=F}
ER <- enricher( over.gss.flanking.peaks.entrez, TERM2GENE = kegggene, pvalueCutoff  = cp.pvalueCutoff, pAdjustMethod = cp.pAdjustMethod, minGSSize = cp.minGSSize, maxGSSize = cp.maxGSSize, qvalueCutoff = cp.qvalueCutoff )

n <- "Curated-KEGG"
if ( !is.null(ER) && nrow(ER) > 0 ) {
    dotplot(ER, title=n)
}
if ( !is.null(ER) && nrow(ER) > 0 ) {
    datatable(as.data.frame(ER), rownames=F)
    write.table( as.data.frame(ER), file=paste0(pipeline.name,"_",n,".tsv") )
}
```

```{r echo=FALSE,dev="png",dpi=300,fig.height=8,fig.width=16,echo=F,message=F}
ER <- enricher( over.gss.flanking.peaks.entrez, TERM2GENE = cpgene, pvalueCutoff  = cp.pvalueCutoff, pAdjustMethod = cp.pAdjustMethod, minGSSize = cp.minGSSize, maxGSSize = cp.maxGSSize, qvalueCutoff = cp.qvalueCutoff )

n <- "ALL-Curated-Canonical-Pathways"
if ( !is.null(ER) && nrow(ER) > 0 ) {
    dotplot(ER, title=n)
}
if ( !is.null(ER) && nrow(ER) > 0 ) {
    datatable(as.data.frame(ER), rownames=F)
    write.table( as.data.frame(ER), file=paste0(pipeline.name,"_",n,".tsv") )
}
```

